src/mscorlib/shared/System/Threading/ExecutionContext.cs

```diff
@@ -130,126 +108,164 @@ public static bool IsFlowSuppressed()
                         {
-                            object currentValue;
-                            current.m_localValues.TryGetValue(local, out currentValue);
+                            next.m_localValues.TryGetValue(local, out object currentValue);
 
                             if (previousValue != currentValue)
+                            {
                                 local.OnValueChanged(previousValue, currentValue, true);
+                            }
                         }
                     }
                 }
-                catch (Exception ex)
-                {
-                    Environment.FailFast(
-                        SR.ExecutionContext_ExceptionInAsyncLocalNotification,
-                        ex);
-                }
+            }
+            catch (Exception ex)
+            {
+                Environment.FailFast(
+                    SR.ExecutionContext_ExceptionInAsyncLocalNotification,
+                    ex);
             }
         }
 
         internal static object GetLocalValue(IAsyncLocal local)
         {
             ExecutionContext current = Thread.CurrentThread.ExecutionContext;
             if (current == null)
+            {
                 return null;
+            }
 
-            object value;
-            current.m_localValues.TryGetValue(local, out value);
+            current.m_localValues.TryGetValue(local, out object value);
             return value;
         }
 
         internal static void SetLocalValue(IAsyncLocal local, object newValue, bool needChangeNotifications)
         {
-            ExecutionContext current = Thread.CurrentThread.ExecutionContext ?? ExecutionContext.Default;
+            ExecutionContext current = Thread.CurrentThread.ExecutionContext;
 
-            object previousValue;
-            bool hadPreviousValue = current.m_localValues.TryGetValue(local, out previousValue);
+            object previousValue = null;
+            bool hadPreviousValue = false;
+            if (current != null)
+            {
+                hadPreviousValue = current.m_localValues.TryGetValue(local, out previousValue);
+            }
 
             if (previousValue == newValue)
+            {
                 return;
+            }
 
-            IAsyncLocalValueMap newValues = current.m_localValues.Set(local, newValue);
+            IAsyncLocal[] newChangeNotifications = null;
+            IAsyncLocalValueMap newValues;
+            bool isFlowSuppressed = false;
+            if (current != null)
+            {
+                isFlowSuppressed = current.m_isFlowSuppressed;
+                newValues = current.m_localValues.Set(local, newValue);
+                newChangeNotifications = current.m_localChangeNotifications;
+            }
+            else
+            {
+                newValues = AsyncLocalValueMap.Empty.Set(local, newValue);;
+                newChangeNotifications = Array.Empty<IAsyncLocal>();
+            }
```

Comment:
Have you measured the cost of setting async locals with this change?

Reply:
Wasn't entirely happy with this; will see if can improve

Reply:
Is better now